pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,850 @@
+	pod, ok := obj.(*kapi.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("couldn't get object from tombstone: %+v", obj))
+			return
+		}
+		pod, ok = tombstone.Obj.(*kapi.Pod)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("tombstone contained object that is not a pod: %+v", obj))
+			return
+		}
+	}
+	if isBuildPod(pod) {
+		bc.enqueueBuildForPod(pod)
+	}
+}
+
+// buildAdded is called by the build informer event handler whenever a build
+// is created
+func (bc *BuildController) buildAdded(obj interface{}) {
+	build := obj.(*buildapi.Build)
+	bc.enqueueBuild(build)
+}
+
+// buildUpdated is called by the build informer event handler whenever a build
+// is updated or there is a relist of builds
+func (bc *BuildController) buildUpdated(old, cur interface{}) {
+	build := cur.(*buildapi.Build)
+	bc.enqueueBuild(build)
+}
+
+// enqueueBuild adds the given build to the queue.
+func (bc *BuildController) enqueueBuild(build *buildapi.Build) {
+	key, err := kcontroller.KeyFunc(build)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for build %#v: %v", build, err))
+		return
+	}
+	bc.queue.Add(key)
+}
+
+// enqueueBuildForPod adds the build corresponding to the given pod to the controller
+// queue. If a build is not found for the pod, then an error is logged.
+func (bc *BuildController) enqueueBuildForPod(pod *kapi.Pod) {
+	buildName := buildutil.GetBuildName(pod)
+	build, err := bc.buildStore.Builds(pod.Namespace).Get(buildName)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't retrieve build for pod %s/%s: %v", pod.Namespace, pod.Name, err))
+		return
+	}
+	bc.enqueueBuild(build)
+}
+
+// handleError is called by the main work loop to check the return of the handleBuild or handlePodDeleted
+// calls. If an error occurred, then the key is re-added to the queue unless it has been retried too many
+// times.
+func (bc *BuildController) handleError(err error, key interface{}) {
+	if err == nil {
+		bc.queue.Forget(key)
+		return
+	}
+
+	if strategy.IsFatal(err) {
+		glog.V(2).Infof("Will not retry fatal error for key %v: %v", key, err)
+		bc.queue.Forget(key)
+		return
+	}
+
+	if bc.queue.NumRequeues(key) < maxRetries {
+		glog.V(4).Infof("Retrying key %v: %v", key, err)
+		bc.queue.AddRateLimited(key)
+		return
+	}
+
+	glog.V(2).Infof("Giving up retrying %v: %v", key, err)
+	bc.queue.Forget(key)
+}
+
+// isBuildPod returns true if the given pod is a build pod
+func isBuildPod(pod *kapi.Pod) bool {
+	return len(buildutil.GetBuildName(pod)) > 0
+}
+
+// buildDesc is a utility to format the namespace/name and phase of a build
+// for errors and logging
+func buildDesc(build *buildapi.Build) string {
+	return fmt.Sprintf("%s/%s (%s)", build.Namespace, build.Name, build.Status.Phase)
+}
+
+// podDeletedUpdate returns a buildUpdate object to apply to a build that has had
+// its pod deleted
+func podDeletedUpdate() *buildUpdate {
```

Comment:
just a thought - is it worth calling this transitionToPhaseError() and accept the reason and message as parameters, then use this as a helper in the places where we transition to error?  It would be clearer than calling it podDeletedUpdate.